### ISSUE TITLE
out of memory

### ISSUE THREAD

**Issue by lastmjs** [sql-js/sql.js#568]
First off, my JavaScript environment is very custom and different than the web or Node.js. We are running in a wasm32-wasi environment with [QuickJS](https://github.com/bellard/quickjs), using [wasmedge-quickjs](https://github.com/second-state/wasmedge-quickjs), and [Wasmtime](https://github.com/bytecodealliance/wasmtime) on the [Internet Computer](https://internetcomputer.org/).

That being said, the strange thing is that I had this working about a week ago, but now it doesn't work. We've been doing a lot of low-level changes to our JS environment (we're working at the engine/platform level), but I just don't understand what could have caused this.

The simple problem is that we get an error just trying to initialize the database: `out of memory`

Here's our code:

```typescript
import { Server } from 'azle';
import express from 'express';
import initSqlJs from 'sql.js/dist/sql-asm-debug.js';

export default Server(async () => {
    try {
        const SQL = await initSqlJs({
            locateFile: (filename) => `/dist/${filename}`,
            TOTAL_STACK: 4_000_000_000,
            TOTAL_MEMORY: 4_000_000_000,
            FAST_MEMORY: 4_000_000_000
        });

        const db = new SQL.Database();

        db.run(`
            CREATE TABLE users
                (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    username TEXT NOT NULL UNIQUE
                );
        `);
    } catch (error) {
        console.log('there was an error');
        console.log(error.message);
        console.log(error.stack);
        console.log(error);
    }

    const app = express();

    app.get('/', (req, res) => {
        res.send('Hello world');
    });

    return app.listen();
});
```

We have to use `asm` as our environment does not have a Wasm interpreter. I've tried `sql-asm-memory-growth.js` as well and I get the same error. I've even gone back multiple versions to maybe 1.7.0 and the error is always there.

Any ideas on what could cause this? I've also tried adding:

```
            TOTAL_STACK: 4_000_000_000,
            TOTAL_MEMORY: 4_000_000_000,
            FAST_MEMORY: 4_000_000_000
```

to the initSqlJs options.

**Comment by lovasoa** [comment]
Do you have an easy and minimal way to reproduce your issue?

@kripken , any idea?

**Comment by lovasoa** [comment]
Are you sure `/dist/${filename}` is what you need? Did you check if the files were loaded correctly in your custom environment.

**Comment by lovasoa** [comment]
Compiling sqlite to js  and quickjs to wasm strikes me as bizarre and inefficient. Why not compile sqlite to wasm directly if you have access to a wasm runtime ?

**Comment by lastmjs** [comment]
> Do you have an easy and minimal way to reproduce your issue?
> 
> @kripken , any idea?

I could get you a reproduction maybe after our next release, but it's not going to be "easy" because you would have to setup our dev environment.

**Comment by lastmjs** [comment]
> Are you sure `/dist/${filename}` is what you need? Did you check if the files were loaded correctly in your custom environment.

I don't think this matters in our case at all since we are using `sql-asm.js`, am I correct? I've tried changing that path and even removing it entirely and the behavior is the same.

**Comment by lastmjs** [comment]
> Compiling sqlite to js and quickjs to wasm to run one on the other strikes me as bizarre and inefficient. Why not compile sqlite to wasm directly if you have access to a wasm runtime ?

We're basically building a Node/Deno for the Internet Computer, and we are trying to allow developers to npm install any package and just use it. So we would like SQL.js to work without modifications.

**Comment by kripken** [comment]
> @kripken , any idea?

Sorry, no - an internal out of memory error is too general.

I'd recommend the usual ideas, though, as before:

* Try a debug build (`-sASSERTIONS`)
* Try a sanitizers build (say, `-fsanitize=undefined,address`)
* [Bisect](https://emscripten.org/docs/contributing/developers_guide.html#bisecting) on emscripten releases

**Comment by lovasoa** [comment]
@lastmjs , what does the debug build output for you ?

**Comment by linhttbk97** [comment]
I got the same issue when upgrade from 1.6.0 -> 1.8.0 only on safari 15.6.1 
on chrome, the issue is gone

**Comment by lastmjs** [comment]
I think I tracked down the source of the issue, though I don't fully understand yet, my understanding is deepening. This seems to have to do with the order of some of our imports, particularly our importing of a polyfill of TextEncoding and TextDecoding. It's relatively project-specific, so not sure if this has anything to do with sql.js itself.

It would be nice if the error could be better...I'm still investigating. I was able to reproduce and fix is the good part.
